### ISSUE TITLE
Mean and max slow

### ISSUE THREAD

**Issue by yaysummeriscoming** [pytorch/pytorch#14615]
## 🐛 Bug

I've recently been optimising some of my code & have found that global average and max pooling seem to be very slow.  Furthermore, there seems to be a great difference in speed between different implementations

## To Reproduce

nb_cycles = 1000
x = torch.FloatTensor(np.random.rand(512, 128, 28, 28)).cuda()
conv = nn.Conv2d(in_channels=128, out_channels=128, kernel_size=1, bias=False).cuda()

with torch.no_grad():
    stt_time = time()
    for _ in range(nb_cycles):
        conv_out = conv(x)
    torch.cuda.synchronize()
    print('Conv only time is: %0.2f s' % (time() - stt_time))

    # Original
    stt_time = time()
    for _ in range(nb_cycles):
        x_mean = F.avg_pool2d(x, kernel_size=28)
    torch.cuda.synchronize()
    print('F.avg_pool2d time is: %0.2f s' % (time() - stt_time))

    stt_time = time()
    for _ in range(nb_cycles):
        x_max = F.max_pool2d(x, kernel_size=28)
    torch.cuda.synchronize()
    print('F.max_pool2d time is: %0.2f s' % (time() - stt_time))

    # This takes about the same amount of time
    stt_time = time()
    for _ in range(nb_cycles):
        x_mean2 = x.mean(-1)
        x_mean2 = x_mean2.mean(-1)
    torch.cuda.synchronize()
    print('2 stage mean time is: %0.2f s' % (time() - stt_time))

    stt_time = time()
    for _ in range(nb_cycles):
        x_max2 = x.max(-1)[0]
        x_max2 = x_max2.max(-1)[0]
    torch.cuda.synchronize()
    print('2 stage max time is: %0.2f s' % (time() - stt_time))

    # This does the same, but is MUCH faster!!!
    stt_time = time()
    for _ in range(nb_cycles):
        x_flattened = x.view(*(x.shape[:-2]), -1)
        x_mean3 = x_flattened.mean(-1)
    torch.cuda.synchronize()
    print('1 stage mean time is: %0.2f s' % (time() - stt_time))

    stt_time = time()
    for _ in range(nb_cycles):
        x_flattened = x.view(*(x.shape[:-2]), -1)
        x_max3 = x_flattened.max(-1)[0]
    torch.cuda.synchronize()
    print('1 stage max time is: %0.2f s' % (time() - stt_time))

    # Reference global mean/amx
    stt_time = time()
    for _ in range(nb_cycles):
        x_mean4 = x.mean()
    torch.cuda.synchronize()
    print('1 stage global mean time is: %0.2f s' % (time() - stt_time))

    stt_time = time()
    for _ in range(nb_cycles):
        x_max4 = x.max()
    torch.cuda.synchronize()
    print('1 stage global max time is: %0.2f s' % (time() - stt_time))

# Output
Conv only time is: 1.02 s
F.avg_pool2d time is: 1.15 s
F.max_pool2d time is: 3.70 s
2 stage mean time is: 0.68 s
2 stage max time is: 0.56 s
1 stage mean time is: 0.24 s
1 stage max time is: 0.53 s
1 stage global mean time is: 0.25 s
1 stage global max time is: 0.29 s

## Expected behavior

I've added in a convolution operation for reference.  It seems weird to me that F.avg_pool2d / F.max_pool2d take longer than the convolution does and much longer than the other implementations?  Also, I would've expected the 1 stage mean/max operations to take negligible time compared to the convolution.  I imagine that this is because the operation is memory bandwidth limited, as it still needs to access every element?

## Environment

Collecting environment information...
PyTorch version: 1.0.0a0+166ee86
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 16.04.5 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: Tesla V100-SXM2-16GB
Nvidia driver version: 410.72
cuDNN version: Probably one of the following:
/usr/local/cuda-10.0/targets/x86_64-linux/lib/libcudnn.so.7.4.1
/usr/local/cuda-10.0/targets/x86_64-linux/lib/libcudnn_static.a

Versions of relevant libraries:
[pip] Could not collect
[conda] Could not collect

**Comment by albanD** [comment]
Hi,

I think all of these are expected results from our point of view:
- pooling layers are usually used with kernels of very small size (2 or 4 in general) and so are optimized to behave well for such kernels. In your test, using kernels of size 28x28 leads to less performances.
- 2 stage mean/max is expected to be slower than 1 stage ones as only one kernel is launched instead of 2. Moreover more work can be done in parallel and so the gpu is better used in that case.
- global reduction is similar as 1 stage one as it is doing very similar work: launch one kernel that reads each element once and accumulate.
- mean and max have slightly different behaviors as they use different kernels (for handling contiguous and innermost dimensions respectively) and so have different scaling.
- Convolution is very well optimized (and many algorithms are available and the best one is picked). Using a kernel 1x1 will give a quite simple operation to be done and so I'm not surprised that it is "only" slightly slower than reduction operations (when done on the whole tensor).

**Comment by yaysummeriscoming** [comment]
Ok, thanks for clearing that up.  Just to convince myself I also ran some tests for batch norm & relu, which I assume also have highly optimised implementations:

nb_cycles = 1000
x = torch.FloatTensor(np.random.rand(512, 128, 28, 28)).cuda()
conv = nn.Conv2d(in_channels=128, out_channels=128, kernel_size=1, bias=False).cuda()
bn = nn.BatchNorm2d(num_features=128).cuda()
relu = nn.ReLU(inplace=True).cuda()

with torch.no_grad():
    stt_time = time()
    for _ in range(nb_cycles):
        conv_out = conv(x)
    torch.cuda.synchronize()
    print('Conv only time is: %0.2f s' % (time() - stt_time))

    stt_time = time()
    for _ in range(nb_cycles):
        bn_out = bn(x)
    torch.cuda.synchronize()
    print('Batch norm only time is: %0.2f s' % (time() - stt_time))

    stt_time = time()
    for _ in range(nb_cycles):
        relu_out = relu(x)
    torch.cuda.synchronize()
    print('Relu only time is: %0.2f s' % (time() - stt_time))

Conv only time is: 2.50 s
Batch norm only time is: 0.82 s
Relu only time is: 0.52 s

With this in mind, I would argue that F.avg_pool2d/F.max_pool2d should use a special faster implementation for global pooling?  Reason being that this seems to be the most widely used global pooling method, such as in the top end of the reference models of torchvision.  

Additionally I don't quite follow your point about slightly different implementations for 1 stage/tensor mean & max?  I could understand 10% or 20% difference in runtime, but in my test 1 stage global max pooling took twice as long as 1 stage global average pooling?

**Comment by jjsjann123** [comment]
Looking at your number, global mean/max seems to have similar performance.

Reading from the context I think you are referring to the performance differences between 1 stage mean & max.
As a simple reduction kernel is used in both cases, which is totally bounded by I/O bandwidth.
Different implementation shouldn't have much impact on the math, since there's not much going on.
However, different launch configurations could have a huge impact on how efficiently GPU bandwidth is being utilized -> exposing enough parallelism work that could be put on the card to fully exploit I/O bandwidth.

**Comment by yaysummeriscoming** [comment]
Correct, I'm referring to 1 stage mean & max.  I'm not following regarding different launch configurations?  These are elementary operations, so I can't understand how they could have such a different implementation/memory access pattern?

**Comment by ngimel** [comment]
`mean` goes through recently added reduction kernels in TensorIterator. It is not without problems, as we've discussed in #12115, but it is fast. `max` uses older kernels in THC. At some point all THC reductions should be moved to TensorIterator reductions, but this future has not arrived yet (PRs welcome).

**Comment by jjsjann123** [comment]
To answer your question earlier:
It's not the memory access pattern. It's launch configurations that could have a big impact on the performance: https://github.com/pytorch/pytorch/blob/65c6cd121eddb35a277fec96285457ad8003a550/aten/src/ATen/native/cuda/Reduce.cuh#L444-L453
Given the case when work per-thread is large and output is small, a naive reduction implementation would take 1 cta per output element. That would not expose enough parallelism to fully utilize the memory bandwidth.

**Comment by yaysummeriscoming** [comment]
@ngimel and @jjsjann123 thanks for the detailed explanations, I'll close the issue now.  I'm a bit hesitant touching source code given that I'm quite new to pytorch, could I be of some assistance with docs & testing?
